### PR TITLE
Fix map_file on Windows by simplifying this repo

### PR DESCRIPTION
**Pull request by aantron**
`Mmap.V1.map_file` [does not work on Windows](https://ci.appveyor.com/project/aantron/lwt/builds/23560678#L471):

```
Failures in test suite 'lwt_bytes':
Test 'map_file' raised 'Invalid_argument("Unix.map_file: not supported")'
        main alias test/unix/runtest (exit 1)
```

By contrast, `Unix.map_file` [works fine](https://ci.appveyor.com/project/aantron/lwt/builds/23566580). `Bigarray.Genarray.map_file` also works fine.

Rather than try to fix this by editing the existing imported code, this PR replaces the entire implementation of the library with a simple shim to either [`Unix.map_file`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Unix.html#VALmap_file) or [`Bigarray.Genarray.map_file`](http://caml.inria.fr/pub/docs/manual-ocaml-4.02/libref/Bigarray.Genarray.html#VALmap_file), based on a compiler version check for 4.06.0. In particular, all the C code is now deleted.

The shim should be easier to maintain.

The implementation selection code is based on [janestreet/result](https://github.com/janestreet/result).

I think we can also replace the license at this point.

The shim [passes Lwt's tests on Windows](https://ci.appveyor.com/project/aantron/lwt/builds/23568058#L298).

### CODE REVIEW

**Comment by aantron** [comment]
Regarding the second commit, I would actually prefer to leak the `Sys_error`, as it would leave the description intact. For the particular case of Lwt, it also is an already-expected behavior.

**Comment by aantron** [comment]
That is, leak the `Sys_error` when it cannot be translated.

**Comment by avsm** [comment]
I'm in favour of this approach, since in the longer term it means we use the upstream implementation post 4.07.

**Comment by avsm** [comment]
If it's expected behaviour for Lwt, then I'm ok with the `Sys_error` too.

**Comment by ghost** [comment]
That seems fine to me, however just to make sure everybody understands the consequences: when a new feature or fix will be provided by `Unix.map_file`, it will not be possible to backport it.

Are you happy with this?

**Comment by aantron** [comment]
I am fine with this for Lwt, because the long-term plan is to switch to `Unix.map_file` anyway once < 4.06 support is dropped, which means that we will be directly relying on each compiler version's implementation of it anyway.

Let me push one more commit to leak `Sys_error` when it cannot be translated on < 4.06, as discussed with @avsm.

File: src/mmap_bigarray.ml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    try Bigarray.Genarray.map_file fd ?pos kind layout shared dims
+    with Sys_error description as exn ->
+      try
+        let error_code =
+          List.find
+            (fun errno -> Unix.error_message errno = description) errno_codes
+        in
+        raise (Unix.Unix_error (error_code, "Unix.map_file", ""))
+      with Not_found ->
+        raise exn
```
Comment (ghost) [inline src/mmap_bigarray.ml:84]: Maybe we should use `raise_with_backtrace`?

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/mmap_bigarray.ml (hunk 1 above)
Comment (aantron) [reply src/mmap_bigarray.ml:84]: AFAIK `raise_with_backtrace` requires 4.05.

**Comment by ghost** [comment]
BTW, instead of relying of compile time selection of code, you can use the following trick we use in dune:

```
include struct
  [@@@warning "-32-3"]
  let map_file = <shim implementation>
end

open Unix

let map_file = map_file
```

With newer versions, the shim implementation will be ignored.

File: src/mmap_bigarray.ml (hunk 1 above)
Comment (ghost) [reply src/mmap_bigarray.ml:84]: Indeed. I suggest to use the trick I mentioned in the main thread

**Comment by aantron** [comment]
In this case, won't `let map_file = <shim implementation>` fail to compile on new compiler versions, due to the lack of `Bigarray.Genarray.map_file`?

File: src/mmap_bigarray.ml (hunk 1 above)
Comment (aantron) [reply src/mmap_bigarray.ml:84]: I can do it, but it seems not worth the benefit of having an improved backtrace on only one old compiler version. We select a different implementation for >= 4.06, and `raise_with_backtrace` will be a shim for `raise` on < 4.05.

Comment (aantron) [reply src/mmap_bigarray.ml:84]: ...especially considering how minimal the difference between the stack traces will be.

**Comment by ghost** [comment]
Yh, indeed that doesn't work well when a function is deleted.

File: src/mmap_bigarray.ml (hunk 1 above)
Comment (ghost) [reply src/mmap_bigarray.ml:84]: Alright, I accept that

**Comment by ghost** [comment]
Thanks for this PR

**Comment by aantron** [comment]
Thank you!

**Comment by avsm** [comment]
@diml wrote:
> That seems fine to me, however just to make sure everybody understands the consequences: when a new feature or fix will be provided by Unix.map_file, it will not be possible to backport it.

Likewise I am fine with that. When OCaml 4.09 is released, depending on 4.07+ as a minimum for most Mirage uses seems reasonable as a "two compiler release window".

**Pull request merged**